### PR TITLE
[WIP] Generalized planning

### PR DESCRIPTION
**Pull request by renato-scaroni**
Just beginning generalized planning. First implementation working relatively well.

### CODE REVIEW

File: Environments.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+#%%
+#################################
+# Environments definition
+#################################
+
+class Navigation():
+
+    def __init__(self, **kwargs):
+        self.__dict__.update(kwargs)
```
Comment (angelolovatto) [inline Environments.py]: Is this a way to put extra arbitrary properties in the instance? I'm not sure this is the most 'pythonic' way to do so. Why not, for example, use a `config` dict  attribute of the instance which can hold arbitrary values?

File: Environments.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        # grid constants
+        self.__size = torch.tensor (self.size, dtype=torch.float32)
```
Comment (angelolovatto) [inline Environments.py]: I'm not familiar with the double leading underscore syntax. Is it supposed to signal that an attribute is 'private'? In that case, I would recommend using a single leading underscore, which is the convention used in Python since PEP8. 

Besides, no user-defined attribute is really private in Python; the double underscore only changes the final name that attribute gets in the instance's dict. See section 3 of [this](https://dbader.org/blog/meaning-of-underscores-in-python) tutorial.

File: Environments.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    def state_size(self):
+        return self.ndim
+
+    @property
+    def action_size(self):
+        return self.ndim
+
+    def transition(self, state, action):
+        # distance to center of grid
+        d = torch.sqrt(torch.sum((state-self.__center)**2))
```
Comment (angelolovatto) [inline Environments.py]: I think you could use [`torch.norm`](https://dbader.org/blog/meaning-of-underscores-in-python) here to streamline the code, but that's just personal preference.

File: Environments.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        next_state = torch.max(torch.min(next_state, self.__size), self.__0_00)
+
+        return next_state
+
+    def reward(self, state, action):
+        # norm L-1 (manhattan distance)
+        # return -torch.sqrt(torch.sum(state - self.__goal))
+
+        # norm L-2 (euclidean distance)
+        return -torch.sqrt(torch.sum((state - self.__goal)**2))
```
Comment (angelolovatto) [inline Environments.py]: Using `torch.norm` here would also allow the user to parametrize the distance used in reward function (by setting `p=1` or `p=2`)

File: Models.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        self.sequential_model = sequential_model
+
+    def train (self, epochs, save_snapshots=True):
+        losses = []
+        total_rewards = []
+        training_snapshots = []
+        self.sequential_model.zero_grad()
+        for e in range(epochs):
+            # beaurocracy
+            self.sequential_optimizer.zero_grad()
```
Comment (angelolovatto) [inline Models.py]: Maybe I'm missing something, but why are there two `zero_grad()`s in succession here? The optimizer already zeroes the grads of the model, as far as I can tell.

File: torch_planning.py [hunk 6]
```diff
@@ -73,6 +73,7 @@ def transition(self, state, action):
 
         # next position
         next_state = state + deceleration * action.tanh()
+        # next_state = state + deceleration * action
```
Comment (angelolovatto) [inline torch_planning.py]: If everything is working as intended, I think this could be safely removed to reduce clutter. Same applies to some other comments in this file.
